fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/SingleSidePredicate.java

```diff
@@ -0,0 +1,103 @@
+     *           /           \             ====>                   /        \
+     *          /             \                                   /          \
+     *  olapScan(t1)     olapScan(t2)            project(t1.a + 1 as c)  project(t2.b + 2 as d)
+     *                                                        |                   |
+     *                                                        |                   |
+     *                                                        |                   |
+     *                                                        |                   |
+     *                                                     olapScan(t1)        olapScan(t2)
+     *TODO: now t1.a + t2.a = t1.b is not in hashJoinConjuncts. The rule will not handle it.
+     */
+    @Override
+    public Rule build() {
+        return logicalJoin()
+                .when(join -> join.getHashJoinConjuncts().stream().anyMatch(expr ->
+                        expr.children().stream().anyMatch(expr1 -> !(expr1 instanceof Slot))))
+                .then(join -> {
+                    List<List<Expression>> exprsOfJoinRelation =
+                            Lists.newArrayList(Lists.newArrayList(), Lists.newArrayList());
+                    Map<Expression, Alias> exprMap = Maps.newHashMap();
+                    Plan left = join.left();
+                    join.getHashJoinConjuncts().forEach(conjunct -> {
+                        Preconditions.checkArgument(conjunct instanceof EqualTo);
+                        Expression[] exprs = conjunct.children().toArray(new Expression[0]);
+                        int tag = checkIfSwap(exprs[0], left);
+                        exprsOfJoinRelation.get(0).add(exprs[tag]);
+                        exprsOfJoinRelation.get(1).add(exprs[tag ^ 1]);
+                        Arrays.stream(exprs).sequential().forEach(expr ->
+                                exprMap.put(expr, new Alias(expr, "expr_" + expr.hashCode())));
+                    });
+                    LogicalJoin<Plan, Plan> join1 = join.withHashJoinConjuncts(join.getHashJoinConjuncts()
+                            .stream().map(equalTo -> equalTo.withChildren(equalTo.children()
+                                    .stream().map(expr -> exprMap.get(expr).toSlot())
+                                    .collect(Collectors.toList())))
+                            .collect(Collectors.toList()));
+                    Iterator<List<Expression>> iter = exprsOfJoinRelation.iterator();
+                    return join1.withChildren(join1.children().stream().map(
+                            plan -> new LogicalProject<>(new ImmutableList.Builder<NamedExpression>()
+                                    .addAll(iter.next().stream().map(expr -> exprMap.get(expr))
+                                            .collect(Collectors.toList()))
+                                    .addAll(plan.getOutput())
+                                    .build(), plan))
```

Comment:
Is it possible produce redundant slot here, such as slots those used only in the hash join expressions